File: miami/csa174/assignments/solutions/program09/Person.java
/**
 * Class that represents a person who is walking on a grid and can go
 * 8 directions (N,S,E,W,NE,NW,SE,SW) in integer steps
 * @author Mike Helmick
 */
public class Person {

	private int xPos = 0;
	private int yPos = 0;
	private int steps = 0;
	
	// directions
	public static final double NE = Math.PI / 4.0;
	public static final double SE = NE + Math.PI / 2.0;
	public static final double SW = SE + Math.PI / 2.0;
	public static final double NW = SW + Math.PI / 2.0;
	
	public Person() {
		// nothing to do
	}

	public int getSteps() {
		return steps;
	}

	public int getXPos() {
		return xPos;
	}

	public int getYPos() {
		return yPos;
	}
	
	/**
	 * Moves the person the requested number of steps, in the sepcified direction
	 * 
	 * @param direction - one of N,S,E,W,NE,NW,SE,SW
	 * @param steps - positive number
	 * @return true if the person was moved, false if parameters are invalid
	 */
	public boolean move( String direction, int steps ) {
		if ( !(validDirection(direction) && validSteps(steps) ) ) {
			return false;
		}
		// needed to accept any direction
		direction = direction.toLowerCase();
		
		this.steps += steps;
		
		if ( direction.equals("n") ) {
			yPos += steps;
		} else if ( direction.equals("s") ) {
			yPos -= steps;
		} else if ( direction.equals("e") ) {
			xPos += steps;
		} else if ( direction.equals("w") ) {
			xPos -= steps;
		} else if ( direction.equals("ne") ) {
			int xc = (int) Math.ceil( steps * Math.cos( NE ) );
			int yc = (int) Math.ceil( steps * Math.sin( NE ) );
			xPos += xc;
			yPos += yc;
			
		} else if ( direction.equals("nw") ) {
			int xc = (int) Math.ceil( steps * Math.cos( NW ) );
			int yc = (int) Math.ceil( steps * Math.sin( NW ) );
			xPos += xc;
			yPos += yc;
			
		} else if ( direction.equals("se") ) {
			int xc = (int) Math.ceil( steps * Math.cos( SE ) );
			int yc = (int) Math.ceil( steps * Math.sin( SE ) );
			xPos += xc;
			yPos += yc;
			
		} else if ( direction.equals("sw") ) {
			int xc = (int) Math.ceil( steps * Math.cos( SW) );
			int yc = (int) Math.ceil( steps * Math.sin( SW ) );
			xPos += xc;
			yPos += yc;
			
		}
		
		return true;
	}
	
	/**
	 * Returns true if the direction is valid
	 * @param direction direction to check
	 * @return true if valid, false if not
	 */
	public boolean validDirection( final String direction ) {
		final String dirLower = direction.toLowerCase();
		return dirLower.equals("n")  || dirLower.equals("s")  ||
		       dirLower.equals("e")  || dirLower.equals("w")  ||
		   	   dirLower.equals("ne") || dirLower.equals("nw") ||
		   	   dirLower.equals("se") || dirLower.equals("sw");
	}
	
	/**
	 * Returns true if the number of steps is valid (>= 1)
	 * @param steps number of steps to check
	 * @return true if the number of steps is valid, false otherwise
	 */
	public boolean validSteps( final int steps ) {
		return steps >= 1;
	}
	
	public String toString() {
		StringBuilder build = new StringBuilder();
		build.append( "You are now at (" );
		build.append( xPos );
		build.append( "," );
		build.append( yPos );
		build.append( ")\n" );
		build.append( "Steps Taken (" );
		build.append( steps );
		build.append( "): " );
		int intervals = (int) Math.ceil( steps / 10.0 );
		for( int i = 0; i < intervals; i++ ) {
			build.append("*");
		}
		
		return build.toString();
	}
	
}

File: uc/computerScience2/spring2014/materials/labs/lab03/Person.cpp
/**
 * Implementation of the Person class.
 */
#include <cmath>
#include <string>
#include <algorithm>

#include <Person.h>

using namespace std;

const double Person::NE = M_PI / 4.0;
const double Person::SE = 2 * M_PI - M_PI / 4.0;
const double Person::SW = M_PI + M_PI / 4.0;
const double Person::NW = M_PI - M_PI / 4.0;

Person::Person() : xPos(0), yPos(0), totalSteps(0) {
}

Person::~Person() {
  // Nothing to do
}
  
int Person::getSteps() const {
  return totalSteps;
}

int Person::getXPos() const {
  return xPos; 
}

int Person::getYPos() const {
  return yPos;
}

int Person::myRound(double val) {
  if (val > 0) {
    return (int) ceil(val);
  } else if (val < 0) {
    return (int) floor(val);
  }
  return 0;
}
  
bool Person::move(const string &direction, int steps) {
  string lowerDir(direction);
  transform(lowerDir.begin(), lowerDir.end(), lowerDir.begin(), tolower);
  if (!(validDirection(lowerDir) && validSteps(steps))) {
    return false;
  }

  // Total steps always goes up by "steps"
  totalSteps += steps;
  // TODO(you): increment xPos and yPos correctly, depending on the value of "lowerDir"
  // the direction has already been lowercased for you.
  
  if (lowerDir == "n") {
    yPos += steps;
  } else if (lowerDir == "s") {
    yPos -= steps;
  } else if (lowerDir == "e") {
    xPos += steps;
  } else if (lowerDir == "w") {
    xPos -= steps;
  } else if (lowerDir == "ne") {
    int xD = myRound(steps * cos(NE));
    int yD = myRound(steps * sin(NE));
    xPos += xD;
    yPos += yD;
  } else if (lowerDir == "nw") {
    int xD = myRound(steps * cos(NW));
    int yD = myRound(steps * sin(NW));
    xPos += xD;
    yPos += yD;
  } else if (lowerDir == "se") {
    int xD = myRound(steps * cos(SE));
    int yD = myRound(steps * sin(SE));
    xPos += xD;
    yPos += yD;
  } else if (lowerDir == "sw") {
    int xD = myRound(steps * cos(SW));
    int yD = myRound(steps * sin(SW));
    xPos += xD;
    yPos += yD;
  }

  return true;
}

bool Person::validDirection(const string &direction) {
  return direction == "n" || direction == "s"
      || direction == "e" || direction == "w"
      || direction == "ne" || direction == "nw"
      || direction == "se" || direction == "sw";
}

bool Person::validSteps(int steps) {
  return steps > 0;
}
